src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs

```diff
@@ -0,0 +1,2255 @@
+        {
+            string source = @"
+public class C
+{
+    public event System.Action Event
+    {
+        init { }
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (4,32): error CS0065: 'C.Event': event property must have both add and remove accessors
+                //     public event System.Action Event
+                Diagnostic(ErrorCode.ERR_EventNeedsBothAccessors, "Event").WithArguments("C.Event").WithLocation(4, 32),
+                // (6,9): error CS1055: An add or remove accessor expected
+                //         init { }
+                Diagnostic(ErrorCode.ERR_AddOrRemoveExpected, "init").WithLocation(6, 9)
+                );
+        }
+
+        [Fact]
+        public void IndexerWithInitOnly()
+        {
+            string source = @"
+public class C
+{
+    public string this[int i]
+    {
+        init { }
+    }
+    public C()
+    {
+        this[42] = null;
+    }
+    public void M1()
+    {
+        this[43] = null; // 1
+    }
+}
+public class Derived : C
+{
+    public Derived()
+    {
+        this[44] = null;
+    }
+    public void M2()
+    {
+        this[45] = null; // 2
+    }
+}
+public class D
+{
+    void M3(C c2)
+    {
+        _ = new C() { [46] = null };
+        c2[47] = null; // 3
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (14,9): error CS8802: Init-only member 'C.this[int]' can only be assigned from a constructor, object initialization or 'with' expression of that type.
+                //         this[43] = null; // 1
+                Diagnostic(ErrorCode.ERR_AssignmentInitOnly, "this[43]").WithArguments("C.this[int]").WithLocation(14, 9),
+                // (25,9): error CS8802: Init-only member 'C.this[int]' can only be assigned from a constructor, object initialization or 'with' expression of that type.
+                //         this[45] = null; // 2
+                Diagnostic(ErrorCode.ERR_AssignmentInitOnly, "this[45]").WithArguments("C.this[int]").WithLocation(25, 9),
+                // (33,9): error CS8802: Init-only member 'C.this[int]' can only be assigned from a constructor, object initialization or 'with' expression of that type.
+                //         c2[47] = null; // 3
+                Diagnostic(ErrorCode.ERR_AssignmentInitOnly, "c2[47]").WithArguments("C.this[int]").WithLocation(33, 9)
+                );
+        }
+
+        [Fact]
+        public void ReadonlyFields()
+        {
+            string source = @"
+public class C
+{
+    public readonly string field;
+    public C()
+    {
+        field = null;
+    }
+    public void M1()
+    {
+        field = null; // 1
+        _ = new C() { field = null }; // 2
```

Comment:
>_ = new C() { field = null }; [](start = 8, length = 29)

It feels like we should allow this because we are in the type that declares the field and we are in an initialization phase. The spec doesn't allow that at the moment, please put this on the list of open design issues.

Reply:
Took note in test plan

---
In reply to: [415209454](https://github.com/dotnet/roslyn/pull/43275#discussion_r415209454) [](ancestors = 415209454)